File: javascript/polygon.js
/**
 * Crea una única función (importante que sólo sea una) que sea capaz
 * de calcular y retornar el área de un polígono.
 * - La función recibirá por parámetro sólo UN polígono a la vez.
 * - Los polígonos soportados serán Triángulo, Cuadrado y Rectángulo.
 * - Imprime el cálculo del área de un polígono de cada tipo.
 * @param {number} polygon 
 * @returns {number}
 */


// Definicion de interfaz Polygon
class Polygon{
    area(){
        throw new Error('Metodo "area" no implementado aun...')
    }
    printArea(){
        console.log(`El area del poligono es ${this.area()}`);
    }
}

// Definicion de la clase triangle
class Triangle extends Polygon {
    constructor(base, height){
        super();
        this.base = base;
        this. height = height;
    }

    area(){
        return (this.base * this.height) / 2;
    }
}

// Definicion de la clase Rectangle
class Rectangle extends Polygon {
    constructor(length, width){
        super();
        this. length = length;
        this.width = width;
    }

    area(){
        return (this.length * this.width)
    }
}

// Definicion de la clase square
class Square extends Polygon{
    constructor(side){
        super();
        this.side = side
    }

    area(){
        return (this.side * this.side)
    }
}

// Main function
function calculateArea(polygon) {
    polygon.printArea();
    return polygon.area();
}

calculateArea(new Triangle(10, 5));
calculateArea(new Rectangle(10, 5));
calculateArea(new Square(10));

File: python/polygon.py
'''
Crea una única función (importante que sólo sea una) que sea capaz
 * de calcular y retornar el área de un polígono.
 * - La función recibirá por parámetro sólo UN polígono a la vez.
 * - Los polígonos soportados serán Triángulo, Cuadrado y Rectángulo.
 * - Imprime el cálculo del área de un polígono de cada tipo.

 Procedimiento:
 1. Crear interfaz de poligono (Calcular e imprimir)
 2. Crear clases de los diferentes poligonos con su respectiva logica
 3. Crear funcion principal que ejecute los metodos de las clases
 4. Instanciar los tres poligonos con datos 
'''

# Definicion de la interfaz principal
class Polygon:
    def area(self):
        raise Exception('El area del poligono aun no se ha definido...')
    def printArea(self):
        print(f'El area del poligono es { round(self.area()) }')

# Definicion de la clase Tiangle
class Triangle(Polygon):
    def __init__(self, base, height):
        self.base = base
        self.height = height

    def area(self):
        return (self.base * self.height) / 2

# Definicion de clase Square
class Square(Polygon):
    def __init__(self, side):
        self.side = side
    
    def area(self):
        return self.side * self.side

# Definicion de clase rectangle
class Rectangle(Polygon):    
    def __init__(self, length, height):
        self.length = length
        self.height = height

    def area(self):
        return self.height * self.length
    
# Main function
def calulate_area(polygon):
    polygon.area()
    return polygon.printArea()


calulate_area(Triangle(12, 12))
calulate_area(Square(12))
calulate_area(Rectangle(12,12))